modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCacheMapEntry.java

```diff
@@ -4464,24 +4465,34 @@ protected void removeValue() throws IgniteCheckedException {
     }
 
     /** {@inheritDoc} */
-    @Override public void updateIndex(SchemaIndexCacheVisitorClosure clo) throws IgniteCheckedException,
+    @Override public GridQueryTypeDescriptor updateIndex(SchemaIndexCacheVisitorClosure clo) throws IgniteCheckedException,
         GridCacheEntryRemovedException {
         lockEntry();
 
         try {
             if (isInternal())
-                return;
+                return null;
 
             checkObsolete();
 
             CacheDataRow row = cctx.offheap().read(this);
 
-            if (row != null)
+            if (row != null) {
                 clo.apply(row);
+
+                return cctx.kernalContext().query().typeByValue(
```

Comment:
It is a bit worrying that we started doing this for every put. It looks like we at least need to pass `false` in the last parameter.

Reply:
Looking at the code more closely, I think it would be better to call this from processPartition method and not even change updateIndex and processKey. Moreover, I suggest to move all of new code under the new system property - in case this have unnoticed performance impact.